File: app/src/main/java/com/porarri/yamabikochat/data/tools/ToolCallingOrchestrator.kt
package com.porarri.yamabikochat.data.tools

import com.porarri.yamabikochat.data.remote.TokenUsageSnapshot
import com.porarri.yamabikochat.utils.DiagnosticsLogger
import kotlinx.coroutines.ensureActive
import org.json.JSONObject
import kotlin.coroutines.coroutineContext

class ToolCallingOrchestrator(
    private val registry: LocalToolRegistry,
    private val maxRounds: Int = DEFAULT_MAX_ROUNDS
) {
    suspend fun run(
        request: ToolTurnRequest,
        invoke: suspend (ToolTurnRequest, Int) -> ToolTurnResponse,
        onActivitiesChanged: (suspend (List<ToolActivityStep>) -> Unit)? = null,
        onProgressChanged: (suspend (ToolCallingProgress) -> Unit)? = null
    ): ToolCallingOutcome {
        val working = ToolTurnRequest(messages = request.messages.toMutableList())
        val activities = mutableListOf<ToolActivityStep>()
        var sources = emptyList<ToolSource>()
        val seenCalls = mutableSetOf<String>()
        var combinedUsage: TokenUsageSnapshot? = null
        val replayMessages = mutableListOf<ToolTurnMessage>()
        val roundLimit = maxOf(1, maxRounds)

        for (round in 1..roundLimit) {
            coroutineContext.ensureActive()
            var response = invoke(working, round)
            combinedUsage = mergedUsage(combinedUsage, response.usage)
            response = response.copy(usage = combinedUsage)

            if (response.toolCalls.isEmpty()) {
                return ToolCallingOutcome(
                    response = response,
                    activities = activities.toList(),
                    sources = sources,
                    rounds = round,
                    replayMessages = replayMessages.toList()
                )
            }

            val assistantToolMessage = ToolTurnMessage(
                    role = "assistant",
                    content = response.text,
                    reasoningContent = response.reasoningSummary,
                    toolCalls = response.toolCalls
                )
            working.messages.add(assistantToolMessage)
            val completedRoundMessages = mutableListOf(assistantToolMessage)

            for (call in response.toolCalls) {
                val duplicateKey = duplicateKey(call)
                var step = ToolActivityStep.started(call = call, round = round)
                activities.add(step)
                onActivitiesChanged?.invoke(activities.toList())
                onProgressChanged?.invoke(
                    ToolCallingProgress(activities.toList(), replayMessages.toList())
                )

                val result = if (seenCalls.contains(duplicateKey)) {
                    ToolResult(
                        callId = call.id,
                        name = call.name,
                        content = """{"error":"Duplicate tool call suppressed"}""",
                        isError = true
                    )
                } else {
                    seenCalls.add(duplicateKey)
                    registry.execute(call)
                }

                step.finish(with = result)
                activities[activities.lastIndex] = step
                sources = mergedSources(sources, result.sources)
                onActivitiesChanged?.invoke(activities.toList())

                val toolMessage = ToolTurnMessage(
                        role = "tool",
                        content = result.content,
                        toolCallId = result.callId,
                        toolName = result.name,
                        toolResultIsError = result.isError
                    )
                working.messages.add(toolMessage)
                completedRoundMessages.add(toolMessage)
                onProgressChanged?.invoke(
                    ToolCallingProgress(activities.toList(), replayMessages.toList())
                )
            }

            replayMessages.addAll(completedRoundMessages)
            onProgressChanged?.invoke(
                ToolCallingProgress(activities.toList(), replayMessages.toList())
            )

            if (round == roundLimit) {
                val message = "Web検索ツールは最大ラウンド数に達したため停止しました。"
                DiagnosticsLogger.log(
                    "Local tool round limit reached rounds=$roundLimit"
                )
                val finalText = response.text.trim().takeIf { it.isNotEmpty() } ?: message
                return ToolCallingOutcome(
                    response = response.copy(text = finalText, toolCalls = emptyList()),
                    activities = activities.toList(),
                    sources = sources,
                    rounds = round,
                    replayMessages = replayMessages.toList()
                )
            }
        }

        throw WebToolException.ParseFailure("Local tool orchestrator ended unexpectedly")
    }

    companion object {
        const val DEFAULT_MAX_ROUNDS = 15

        private fun duplicateKey(call: ToolCall): String {
            val arguments = runCatching {
                val obj = JSONObject(call.argumentsJSON)
                val keys = obj.keys().asSequence().sorted().toList()
                val normalized = JSONObject()
                for (key in keys) {
                    normalized.put(key, obj.get(key))
                }
                normalized.toString()
            }.getOrDefault(call.argumentsJSON.trim())
            return "${call.name}\n$arguments"
        }

        private fun mergedSources(
            current: List<ToolSource>,
            incoming: List<ToolSource>
        ): List<ToolSource> {
            val result = current.toMutableList()
            val seen = current.map { it.url }.toMutableSet()
            for (source in incoming) {
                if (seen.add(source.url)) {
                    result.add(source)
                }
            }
            return result
        }

        private fun mergedUsage(
            current: TokenUsageSnapshot?,
            incoming: TokenUsageSnapshot?
        ): TokenUsageSnapshot? {
            if (current == null && incoming == null) return null
            fun sum(lhs: Int?, rhs: Int?): Int? {
                if (lhs == null && rhs == null) return null
                return maxOf(0, lhs ?: 0) + maxOf(0, rhs ?: 0)
            }
            val merged = TokenUsageSnapshot(
                inputTokens = sum(current?.inputTokens, incoming?.inputTokens) ?: 0,
                outputTokens = sum(current?.outputTokens, incoming?.outputTokens) ?: 0,
                totalTokens = sum(current?.totalTokens, incoming?.totalTokens) ?: 0,
                reasoningTokens = sum(current?.reasoningTokens, incoming?.reasoningTokens),
                cachedInputTokens = sum(current?.cachedInputTokens, incoming?.cachedInputTokens)
            ).normalized()
            return merged.takeUnless { it.isEmpty() }
        }
    }
}

File: ios/YamabikoChat/Features/Tools/Core/ToolCallingOrchestrator.swift
import Foundation

struct ToolCallingOutcome: Sendable, Equatable {
    var response: ProviderResponse
    var activities: [ToolActivityStep]
    var sources: [ToolSource]
    var rounds: Int
    var replayMessages: [ProviderRequestMessage]
}

struct ToolCallingProgress: Sendable, Equatable {
    var activities: [ToolActivityStep]
    /// Only complete assistant-tool/result rounds are exposed for durable replay.
    var replayMessages: [ProviderRequestMessage]
}

struct ToolCallingOrchestrator: Sendable {
    static let defaultMaxRounds = 15

    let registry: LocalToolRegistry
    var maxRounds: Int = Self.defaultMaxRounds

    func run(
        request initialRequest: ProviderRequest,
        invoke: @Sendable (ProviderRequest, Int) async throws -> ProviderResponse,
        onActivitiesChanged: (@Sendable ([ToolActivityStep]) async -> Void)? = nil,
        onProgressChanged: (@Sendable (ToolCallingProgress) async -> Void)? = nil
    ) async throws -> ToolCallingOutcome {
        var request = initialRequest
        var activities: [ToolActivityStep] = []
        var sources: [ToolSource] = []
        var seenCalls: Set<String> = []
        var combinedUsage: ProviderUsage?
        var replayMessages: [ProviderRequestMessage] = []
        let roundLimit = max(1, maxRounds)

        for round in 1 ... roundLimit {
            try Task.checkCancellation()
            var response = try await invoke(request, round)
            combinedUsage = Self.mergedUsage(combinedUsage, response.usage)
            response.usage = combinedUsage

            guard !response.toolCalls.isEmpty else {
                return ToolCallingOutcome(
                    response: response,
                    activities: activities,
                    sources: sources,
                    rounds: round,
                    replayMessages: replayMessages
                )
            }

            let assistantToolMessage = ProviderRequestMessage(
                    role: "assistant",
                    content: response.text,
                    reasoningContent: response.reasoningSummary,
                    toolCalls: response.toolCalls
                )
            request.messages.append(assistantToolMessage)
            var completedRoundMessages = [assistantToolMessage]

            for call in response.toolCalls {
                let duplicateKey = Self.duplicateKey(for: call)
                var step = ToolActivityStep.started(call: call, round: round)
                activities.append(step)
                await onActivitiesChanged?(activities)
                await onProgressChanged?(ToolCallingProgress(
                    activities: activities,
                    replayMessages: replayMessages
                ))

                let result: ToolResult
                if seenCalls.contains(duplicateKey) {
                    result = ToolResult(
                        callId: call.id,
                        name: call.name,
                        content: #"{"error":"Duplicate tool call suppressed"}"#,
                        isError: true
                    )
                } else {
                    seenCalls.insert(duplicateKey)
                    let toolStartedAtMs = Int64(Date().timeIntervalSince1970 * 1_000)
                    result = await registry.execute(call: call)
                    let toolCompletedAtMs = Int64(Date().timeIntervalSince1970 * 1_000)
                    if let metrics = ProviderMetricsContext.current {
                        metrics.recorder(
                            ConversationExecutionMetric(
                                conversationId: metrics.conversationId,
                                turnId: metrics.turnId,
                                kind: .tool,
                                startedAtMs: toolStartedAtMs,
                                firstTokenAtMs: nil,
                                completedAtMs: toolCompletedAtMs,
                                succeeded: !result.isError,
                                inputTokens: nil,
                                outputTokens: nil,
                                cachedInputTokens: nil,
                                cacheCreationInputTokens: nil
                            )
                        )
                    }
                }

                step.finish(with: result)
                activities[activities.count - 1] = step
                sources = Self.mergedSources(sources, result.sources)

                let toolMessage = ProviderRequestMessage(
                        role: "tool",
                        content: result.content,
                        toolCallId: result.callId,
                        toolName: result.name,
                        toolResultIsError: result.isError
                    )
                request.messages.append(toolMessage)
                completedRoundMessages.append(toolMessage)
                await onActivitiesChanged?(activities)
                await onProgressChanged?(ToolCallingProgress(
                    activities: activities,
                    replayMessages: replayMessages
                ))
            }

            replayMessages.append(contentsOf: completedRoundMessages)
            await onProgressChanged?(ToolCallingProgress(
                activities: activities,
                replayMessages: replayMessages
            ))

            if round == roundLimit {
                let message = L10n.text("Web検索ツールは最大ラウンド数に達したため停止しました。")
                DiagnosticsLogger.log(
                    "Local tool round limit reached",
                    level: .warning,
                    category: .chat,
                    metadata: ["rounds": String(roundLimit)]
                )
                response.text = response.text.trimmedNonEmpty ?? message
                response.toolCalls = []
                return ToolCallingOutcome(
                    response: response,
                    activities: activities,
                    sources: sources,
                    rounds: round,
                    replayMessages: replayMessages
                )
            }
        }

        throw ProviderClientError.parseFailure("Local tool orchestrator ended unexpectedly")
    }

    private static func duplicateKey(for call: ToolCall) -> String {
        if call.name == WebSearchTool.name,
           let data = call.argumentsJSON.data(using: .utf8),
           let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let query = object["query"] as? String {
            let limit = ToolArguments.int(object["max_results"]) ?? DuckDuckGoHTMLEngine.resultLimit
            return "\(call.name)\nquery=\(RelevantPageReader.normalizeSearchQuery(query))\nmax_results=\(limit)"
        }
        let arguments: String
        if let data = call.argumentsJSON.data(using: .utf8),
           let object = try? JSONSerialization.jsonObject(with: data),
           let normalized = try? JSONSerialization.data(withJSONObject: object, options: [.sortedKeys]),
           let json = String(data: normalized, encoding: .utf8) {
            arguments = json
        } else {
            arguments = call.argumentsJSON.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return "\(call.name)\n\(arguments)"
    }

    private static func mergedSources(_ current: [ToolSource], _ incoming: [ToolSource]) -> [ToolSource] {
        var result = current
        var seen = Set(current.map(\.url))
        for source in incoming where seen.insert(source.url).inserted {
            result.append(source)
        }
        return result
    }

    private static func mergedUsage(_ current: ProviderUsage?, _ incoming: ProviderUsage?) -> ProviderUsage? {
        guard current != nil || incoming != nil else { return nil }
        func sum(_ lhs: Int?, _ rhs: Int?) -> Int? {
            guard lhs != nil || rhs != nil else { return nil }
            return max(0, lhs ?? 0) + max(0, rhs ?? 0)
        }
        return ProviderUsage(
            inputTokens: sum(current?.inputTokens, incoming?.inputTokens),
            outputTokens: sum(current?.outputTokens, incoming?.outputTokens),
            totalTokens: sum(current?.totalTokens, incoming?.totalTokens),
            reasoningTokens: sum(current?.reasoningTokens, incoming?.reasoningTokens),
            cachedInputTokens: sum(current?.cachedInputTokens, incoming?.cachedInputTokens),
            cacheCreationInputTokens: sum(
                current?.cacheCreationInputTokens,
                incoming?.cacheCreationInputTokens
            )
        )
        .normalizedNonEmpty()
    }
}
